employees/views.py

```diff
@@ -37,41 +39,80 @@ def query_as_dict(query_set):
 class ReportList(APIView):
     renderer_classes = [renderers.TemplateHTMLRenderer]
     template_name = 'employees/report_list.html'
+    reports_dict = {}
+    project_form = ''
     permission_classes = (
         permissions.IsAuthenticated,
     )
 
     def get_queryset(self):
         return Report.objects.filter(author=self.request.user).order_by('-date', 'project__name')
 
+    def _add_project(self, serializer, project):
+        project.members.add(self.request.user)
+        project.full_clean()
+        project.save()
+        project.refresh_from_db()
+        serializer.fields['project'].initial = project
+
+    def _create_serializer(self):
+        reports_serializer = ReportSerializer(context={'request': self.request}, )
+        reports_serializer.fields['project'].queryset = \
+            Project.objects.filter(
+                members__id=self.request.user.id
+            ).order_by('name')
+        return reports_serializer
+
+    def initial(self, request, *args, **kwargs):
+        super().initial(request, *args, **kwargs)
+        self.reports_dict = query_as_dict(self.get_queryset())
+        self.project_form = ProjectJoinForm(
+            queryset=Project.objects.exclude(members__id=self.request.user.id).order_by('name'),
+        )
+
     def get(self, request):
-        reports_serializer = ReportSerializer(context={'request': request})
-        reports_dict = query_as_dict(self.get_queryset())
         return Response({
-            'serializer': reports_serializer,
-            'reports_dict': reports_dict,
+            'serializer': self._create_serializer(),
+            'reports_dict': self.reports_dict,
             'UI_text': ReportListStrings,
+            'project_form': self.project_form,
         })
 
     def post(self, request):
         reports_serializer = ReportSerializer(data=request.data, context={'request': request})
-        reports_dict = query_as_dict(self.get_queryset())
+
+        if 'join' in request.POST:
```

Comment:
This could be changed to some kind of Enum and reused in place where it is assigned to request. Thanks to this will not stop working if it will be changed in another place

Reply:
You mean the 'join' string? It's an id from HTML template, so I'd say it's name is rather irrelevant, but I could supply some sort of constant for it. The question is whether it's worth the extra effort?

Reply:
Yea, you're right. It won't be necessary :)